feat: Purge `purge` Controller (#3506)
kyma-project/lifecycle-manager: notable-changes/v1.21.0/notable-change.md

> This migration is mandatory. Run the migration script immediately after updating Kyma Lifecycle Manager (KLM) to version 1.21.0 or greater. If you delay, any Kyma custom resource marked for deletion will remain stuck in the deletion process indefinitely. The purge controller that previously removed this finalizer no longer exists in KLM version 1.21.0.
 
## Prerequisites

Before you begin, ensure you have the following:

- `kubectl` configured to point to the KCP cluster.
- `yq` installed on the machine where you run the script.
- KLM updated to version 1.21.0 or greater in the `kcp-system` namespace.

## What's Changed

As of version 1.21.0, the purge controller has been removed from KLM. Previously, KLM added the finalizer `operator.kyma-project.io/purge-finalizer` to every Kyma custom resource. The finalizer coordinated forced cleanup of remote cluster resources during deprovisioning.

With the purge controller removed, nothing removes this finalizer anymore. Any Kyma custom resource that still carries it and enters deletion will block indefinitely. The resource waits for a finalizer owner that no longer exists.

## Procedure

1. Update KLM to version 1.21.0 or greater.

> 2. Immediately after the update, run the [migration script](./remove-purge-finalizer.sh) co-located with this document:

Reviewer: ```suggestion
2. Immediately after the update, run the [migration script](./remove-purge-finalizer.sh).
```

   ```bash
   # Dry run first - shows affected Kyma custom resources without making any changes
   ./remove-purge-finalizer.sh

   # Apply the changes once you have verified the dry-run output
   ./remove-purge-finalizer.sh --execute
   ```

   The script reads the `app.kubernetes.io/version` label of the `klm-controller-manager` Deployment in the `kcp-system` namespace and warns if KLM has not yet been updated to at least 1.21.0. If the version check cannot be completed (for example, the label is missing or not in semver format), the script prompts you to confirm manually before proceeding.

> 3. You must answer with a capital `YES` in order for the script to proceed. Anything other than a `YES` will fail the script.

Reviewer: ```suggestion
3. You must answer with a capital `YES` for the script to proceed. Anything other than`YES` fails the script.
```

<!--
{
  "metadata": {
    "requirement": "MANDATORY",
    "type": "INTERNAL",
    "category": "MIGRATION"
  }
}
-->

# KLM: Remove the Purge Finalizer from All Kyma Custom Resources

> ### Caution:

> > This migration is mandatory. Run the migration script immediately after updating KLM to v1.21.0 or greater. If you delay running the script, any Kyma custom resource that is marked for deletion will be stuck in the deletion process indefinitely, because the purge controller that previously removed this finalizer no longer exists in KLM v1.21.0.

Reviewer: ```suggestion
> This migration is mandatory. Run the migration script immediately after updating Kyma Lifecycle Manager (KLM) to v1.21.0 or greater. If you delay, any Kyma custom resource marked for deletion will remain stuck in the deletion process indefinitely. The purge controller that previously removed this finalizer no longer exists in KLM v1.21.0.
```

The final sentence is approximately 35 words; splitting into two sentences and trimming 'that is' to a participial improves density and readability.

## Prerequisites

> 

Reviewer: ```suggestion

Before you begin, ensure you have the following:

```

- `kubectl` configured to point to the KCP cluster
- `yq` installed on the machine where you run the script

> - KLM updated to v1.21.0 or greater in the `kcp-system` namespace

Reviewer: ```suggestion
- `kubectl` configured to point to the KCP cluster.
- `yq` installed on the machine where you run the script.
- KLM updated to v1.21.0 or greater in the `kcp-system` namespace.
```

## What's Changed

> As of v1.21.0, the purge controller has been removed from KLM. Previously, KLM added the finalizer `operator.kyma-project.io/purge-finalizer` to every Kyma custom resource to coordinate forced cleanup of remote cluster resources during deprovisioning.

Reviewer: ```suggestion
As of version 1.21.0, the purge controller has been removed from KLM. Previously, KLM added the finalizer `operator.kyma-project.io/purge-finalizer` to every Kyma custom resource. The finalizer coordinated forced cleanup of remote cluster resources during deprovisioning.
```

> With the purge controller removed, nothing removes this finalizer anymore. Any Kyma custom resource that still carries it and enters deletion will block indefinitely, waiting for a finalizer owner that no longer exists.

Reviewer: ```suggestion
With the purge controller removed, nothing removes this finalizer anymore. Any Kyma custom resource that still carries it and enters deletion will block indefinitely. The resource waits for a finalizer owner that no longer exists.
```

## Procedure

> 1. Update KLM to v1.21.0 or greater.

Reviewer: ```suggestion
1. Update KLM to version 1.21.0 or greater.
```

2. Immediately after the update, run the migration script from the KLM repository root:

   ```bash
   # Dry run first - shows affected Kyma custom resources without making any changes
   ./scripts/remove-purge-finalizer.sh --min-version v1.21.0

   # Apply the changes once you have verified the dry-run output
   ./scripts/remove-purge-finalizer.sh --min-version v1.21.0 --execute
   ```

>    The script performs a sanity check at startup: it reads the `app.kubernetes.io/version` label of the `controller-manager` Deployment in the `kcp-system` namespace and aborts if KLM has not yet been updated to at least v1.21.0. This prevents accidentally running the script against an old KLM version that would re-add the finalizer.

Reviewer: ```suggestion
   The script performs a sanity check at startup: it reads the `app.kubernetes.io/version` label of the `controller-manager` Deployment in the `kcp-system` namespace. The script stops if KLM has not yet been updated to at least version 1.21.0. This prevents running the script against an old KLM version that would re-add the finalizer.
```

The first sentence is approximately 38 words; splitting at the colon and removing the redundant adverb 'accidentally' improves clarity and reduces density.

Author: This sentence has been adapted to the newest change to the script behaviour.

>    If the deployed version label is not in semver format, the script exits with a warning. In that case, after manually confirming that the deployed version is v1.21.0 or greater, run the script with `--skip-version-check`:

Reviewer: ```suggestion
   If the deployed version label is not in semver format, the script exits with a warning. In that case, manually confirm that the deployed version is 1.21.0 or greater. Then run the script with `--skip-version-check`:
```

Author: This flag has been now removed. Updated the documentation as well.
